internal/services/ingest.go

```diff
@@ -155,6 +169,212 @@ func (m *ingestService) Run(ctx context.Context, startLedger uint32, endLedger u
+		case <-ctx.Done():
+			log.Ctx(ctx).Info("Context cancelled")
+			return fmt.Errorf("context cancelled: %w", ctx.Err())
+		case <-manualTriggerChan:
+			log.Ctx(ctx).Info("Manual trigger received")
+		case <-ticker.C:
+			log.Ctx(ctx).Info("Ticker ticked")
+		}
+
+		// fetch ledgers
+		getLedgersResponse, err := m.fetchNextLedgersBatch(ctx, startLedger)
+		if err != nil {
+			if errors.Is(err, ErrAlreadyInSync) {
+				log.Ctx(ctx).Info("Ingestion is already in sync, will retry in a few moments...")
+				continue
+			}
+			return fmt.Errorf("fetching next ledgers batch: %w", err)
+		}
+
+		totalIngestionStart := time.Now()
+		// process ledgers
+		err = m.processLedgerResponse(ctx, getLedgersResponse)
+		if err != nil {
+			return fmt.Errorf("processing ledger response: %w", err)
+		}
+
+		// update cursor
+		startLedger = getLedgersResponse.Ledgers[len(getLedgersResponse.Ledgers)-1].Sequence
+		if err := m.models.IngestStore.UpdateLatestLedgerSynced(ctx, m.ledgerCursorName, startLedger); err != nil {
+			return fmt.Errorf("updating latest synced ledger: %w", err)
+		}
+		m.metricsService.SetLatestLedgerIngested(float64(getLedgersResponse.LatestLedger))
+		m.metricsService.ObserveIngestionDuration(totalIngestionPrometheusLabel, time.Since(totalIngestionStart).Seconds())
+
+		if len(getLedgersResponse.Ledgers) == getLedgersLimit {
+			ticker.Reset(tickerDuration)
+			manualTriggerChan <- nil
+		}
+	}
+}
+
+// fetchNextLedgersBatch fetches the next batch of ledgers from the RPC service.
+func (m *ingestService) fetchNextLedgersBatch(ctx context.Context, startLedger uint32) (GetLedgersResponse, error) {
+	rpcHealth, err := m.rpcService.GetHealth()
+	if err != nil {
+		return GetLedgersResponse{}, fmt.Errorf("getting rpc health: %w", err)
+	}
+	ledgerSeqRange, inSync := getLedgerSeqRange(rpcHealth.OldestLedger, rpcHealth.LatestLedger, startLedger)
+	log.Ctx(ctx).Debugf("ledgerSeqRange: %+v", ledgerSeqRange)
+	if inSync {
+		return GetLedgersResponse{}, ErrAlreadyInSync
+	}
+
+	getLedgersResponse, err := m.rpcService.GetLedgers(ledgerSeqRange.StartLedger, ledgerSeqRange.Limit)
+	if err != nil {
+		return GetLedgersResponse{}, fmt.Errorf("getting ledgers: %w", err)
+	}
+
+	return getLedgersResponse, nil
+}
+
+type LedgerSeqRange struct {
+	StartLedger uint32
+	Limit       uint32
+}
+
+// getLedgerSeqRange returns a ledger sequence range to ingest. It takes into account:
+// - the ledgers available in the RPC,
+// - the latest ledger synced by the ingestion service,
+// - the max ledger window to ingest.
+//
+// The returned ledger sequence range is inclusive of the start and end ledgers.
+func getLedgerSeqRange(rpcOldestLedger, rpcNewestLedger, latestLedgerSynced uint32) (ledgerRange LedgerSeqRange, inSync bool) {
+	if latestLedgerSynced >= rpcNewestLedger {
+		return LedgerSeqRange{}, true
+	}
+	ledgerRange.StartLedger = max(latestLedgerSynced+1, rpcOldestLedger)
```

Comment:
If we do a `max()` here would that mean if wallet-backend is at 5 and rpc's oldest ledger is at 8, it would skip ingesting 6, 7? If we are ingesting using `GetLedgers` by getting ledgers in bulk, do we want to allow WB  catch up to RPC instead of skipping ledgers?